### PR TITLE
[dotnet] Resource package registration in .NET

### LINKED ISSUE #5800: [codegen/dotnet] Register package/module resources
.NET side of https://github.com/pulumi/pulumi/pull/5806

### PR DESCRIPTION
**Pull request by mikhailshilkov**
Introduce package and module discovery to the .NET SDK. Resolve #5800.

The implementation relies heavily on reflection:

1. Most public package/module related APIs are removed. Technically, it's breaking, but it does not affect anyone.
2. Every provider gets a single class `ResourcePackage` which implements both provider and custom resource creation.
3. At the time of the first module lookup, we scan the list of referenced assemblies and search for all `ResourcePackage` types. This dictionary is built once and then reused.

All downstream providers fail to build because they haven't been updated to use #5817 yet. I did test Random and Azure locally.

### CODE REVIEW

**Comment by mikhailshilkov** [comment]
Okay, I found https://github.com/pulumi/pulumi/pull/5805 so I need to add version comparisons.

Though - how could two versions of the same package be referenced at the same time? What is the exact scenario of version resolution?

File: sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs [hunk 1]
```diff
@@ -1,40 +1,98 @@
...
+                {
+                    if (PossibleMatch(reference))
+                    {
+                        var assembly = Assembly.Load(reference);
+                        assembliesToCheck.Enqueue(assembly);
+                    }
+                }
+            }
+
+            static bool PossibleMatch(AssemblyName? assembly) => assembly != null && !assembly.FullName.StartsWith("System");
```
Comment (justinvp) [inline sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:95]: Nit: `StartsWith` does a culture-sensitive comparison by default. Best to specify an explicit `StringComparison`.

```suggestion
            static bool PossibleMatch(AssemblyName? assembly) => assembly != null && !assembly.FullName.StartsWith("System", StringComparison.Ordinal);
```

File: sdk/dotnet/Pulumi.Tests/Serialization/ResourcePackagesTests.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                throw new System.NotImplementedException();
+            }
+
+            public Resource Construct(string name, string type, string urn)
+            {
+                throw new System.NotImplementedException();
+            }
+        }
+        
+        private class VersionBlankTestPackage : BaseTestPackage
```
Comment (justinvp) [inline sdk/dotnet/Pulumi.Tests/Serialization/ResourcePackagesTests.cs:81]: Why does this have "Blank" in the name when there is a version specified?

File: pkg/codegen/dotnet/gen.go [hunk 3]
```diff
@@ -1021,6 +1021,72 @@ func (mod *modContext) genEnums(w io.Writer, enums []*schema.EnumType) error {
...
+	fmt.Fprintf(w, "namespace %s\n", assemblyName)
+	fmt.Fprintf(w, "{\n")
+
+	fmt.Fprintf(w, "    internal class ResourcePackage : IResourcePackage\n")
+	fmt.Fprintf(w, "    {\n")
+	fmt.Fprintf(w, "        public string Name => \"%s\";\n\n", pkg.Name)
+	fmt.Fprintf(w, "        public string? Version => null;\n\n")
+	fmt.Fprintf(w, "        [Obsolete(\"The overload with 4 parameters is deprecated and will be removed\")]\n")
+	fmt.Fprintf(w, "        public ProviderResource ConstructProvider(string name, string type, System.Collections.Generic.IDictionary<string, object?>? args, string urn)\n")
+	fmt.Fprintf(w, "            => ConstructProvider(name, type, urn);\n\n")
```
Comment (justinvp) [inline pkg/codegen/dotnet/gen.go:1058]: I don't quite follow why we need to keep this.

File: pkg/codegen/dotnet/gen.go [hunk 4]
```diff
@@ -1021,6 +1021,72 @@ func (mod *modContext) genEnums(w io.Writer, enums []*schema.EnumType) error {
...
+	}
+
+	// Open the namespace.
+	fmt.Fprintf(w, "namespace %s\n", assemblyName)
+	fmt.Fprintf(w, "{\n")
+
+	fmt.Fprintf(w, "    internal class ResourcePackage : IResourcePackage\n")
+	fmt.Fprintf(w, "    {\n")
+	fmt.Fprintf(w, "        public string Name => \"%s\";\n\n", pkg.Name)
+	fmt.Fprintf(w, "        public string? Version => null;\n\n")
```
Comment (justinvp) [inline pkg/codegen/dotnet/gen.go:1055]: I added version info to Node.js and Python in https://github.com/pulumi/pulumi/pull/5839. Want to do the same here?

File: sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs [hunk 5]
```diff
@@ -1,40 +1,98 @@
...
+                .Select(c => c.Invoke(new object[0]))
+                .Cast<IResourcePackage>()
+                .ToImmutableList();
         }
-
-        public static void RegisterResourcePackage(string name, string version, IResourcePackage package)
+        
+        // Assemblies are loaded on demand, so it could be that some assemblies aren't yet loaded to the current
+        // app domain at the time of discovery. This method iterates through the list of referenced assemblies
+        // recursively.
```
Comment (justinvp) [inline sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:60]: Will we be able to address this in the future, if needed (e.g. via `AssemblyLoad` event)?

File: sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs [hunk 6]
```diff
@@ -1,40 +1,98 @@
...
+        // Note: If an assembly is referenced but its types are never used anywhere in the program, that reference
+        // will be optimized out and won't appear in the result of the enumeration.
```
Comment (justinvp) [inline sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:62]: How would someone mitigate this in their program if they're running into it?

**Review by justinvp — APPROVED** [review]
Thanks for putting this together, @mikhailshilkov, and apologies for the delay reviewing it! Some comments, but LGTM otherwise.

@pgavlin, any other feedback?

File: sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs (hunk 5 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:60]: This whole loop through `GetReferencedAssemblies` and call `Assembly.Load` below is to work around this problem. I'll extend the comment.

File: sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs (hunk 6 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:62]: They would have to use something from that assembly in their code. This won't happen often in practice, I suspect.

File: sdk/dotnet/Pulumi.Tests/Serialization/ResourcePackagesTests.cs (hunk 2 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi.Tests/Serialization/ResourcePackagesTests.cs:81]: This looks like a left-over, I'll delete this class, sorry for that

File: pkg/codegen/dotnet/gen.go (hunk 3 above)
Comment (mikhailshilkov) [reply pkg/codegen/dotnet/gen.go:1058]: If a provider updates to the latest `pulumi/pulumi` Go package before updating to a published version of NuGet, the generated code won't compile without this shim. We can remove it if we think it's not a problem.

Comment (justinvp) [reply pkg/codegen/dotnet/gen.go:1058]: Ah, ok. Leaving it sounds fine. Let's open an issue to track removing it.

File: sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs [hunk 7]
```diff
@@ -1,40 +1,98 @@
...
 
 namespace Pulumi
 {
     public interface IResourcePackage
     {
-        ProviderResource ConstructProvider(string name, string type, IDictionary<string, object?>? args, string urn);
+        string Name { get; }
+        string? Version { get; }
+        ProviderResource ConstructProvider(string name, string type, string urn);
+        Resource Construct(string name, string type, string urn);
```
Comment (pgavlin) [inline sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:18]: Just a note that coupling these operations--provider and resource construction--constrains Pulumi SDKs s.t. they are delivered in the same assembly. I _think_ that's fine, but I want to call it out explicitly.

Comment (pgavlin) [reply sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:18]: One alternative that we might want to consider is marking resource classes with an attribute that records their type token and then mapping from type token to class. I think that approach is additive, and we could switch to it in the future without any breaks.

**Review by pgavlin — APPROVED** [review]
LGTM! Just a couple notes re: the `IResourcePackage` approach.

File: sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs (hunk 7 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:18]: These are brilliant points. I went ahead and drafted an alternative implementation based on attributes. I fired a separate PR https://github.com/pulumi/pulumi/pull/5875 to be able to discuss both. I find it simpler, more flexible, and more idiomatic. PTAL cc @justinvp

Comment (justinvp) [reply sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs:18]: I also like the idea of using an attribute to record the type token. I'll take a look at the other PR.

**Pull request closed without merging**